[c10d] Make broadcast as a custom op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75924

Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

Commit:
[c10d] Make broadcast as a custom op

Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -186,4 +186,11 @@ void ProcessGroup::init() {
 }
 
+TORCH_LIBRARY(c10d, m) {
+  m.class_<ProcessGroup>("ProcessGroup")
```
If you're still getting import errors you might need to move it into the same file as the operator registration. You require this type registration to run before the operator can be registered. But there's no guaranteed order of execution for C++ static initialization from different modules.

alanwaketan replied:
Good catch.

alanwaketan replied:
It solved the issue. Thanks, Dima!

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

alanwaketan commented on aot.py:
```diff
@@ -0,0 +1,65 @@
+    setup(rank, world_size)
+
+    # Pass on the compiler_fn to the aot_function API
+    aot_print_fn = aot_function(broadcast, fw_compiler=compiler_fn, bw_compiler=compiler_fn)
```
@Chillee I added example of AOT here. And this is the output. Looks like it works just don't know how to deal with the `bw_compiler`:
```
(pytorch37) jwtan@dev-st-p4d24xlarge-2:/fsx/users/jwtan/work/pytorch$ python aot.py 
main process id: 96435
Running basic DDP example on rank 0, and process id: 96478
Running basic DDP example on rank 1, and process id: 96479
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
```

alanwaketan replied:
Looks like AOT doesn't work... Debugging...

alanwaketan replied:
Added a print after __torch_dispatch__ in AOT: https://github.com/pytorch/functorch/blob/main/functorch/_src/python_key.py#L73

Nothing got printed out... So it seems __torch_dispatch__ cannot recognize torch.ops...

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

Commit:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

ezyang committed:
Update on "[c10d] Make broadcast as a custom op"


Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

[ghstack-poisoned]

alanwaketan commented:
#76722 is the formal version of this pull request. Let's close this one.  Thanks Edward for helping me out on the dispatch key stuff.
